### PR TITLE
YTEP 0034: File standard for FITS images in yt

### PR DESCRIPTION
**Pull request by jzuhone**
This YTEP defines a standard for reading and writing FITS images to and from disk using yt.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
As far as I know you're probably the biggest user of this functionality in yt and its original author. I'm not sure who would be a good person to review this. I've given it a brief read-through and don't see any issues but then again I haven't thought too carefully about this problem.

Would it be helpful for us to involve some of the astropy people? Is there anyone else who regularly uses yt to generate synthetic observations that might be a good person to look through this?

**Comment by matthewturk** [comment]
The Trident project folks may have a vested interest in this; perhaps they would be good reviewer candidates.  @chummels @brittonsmith , do you think that you could provide meaningful feedback on this?

I have also read it over and it all seems quite reasonable to me -- it seems very clearly to serve the purpose of distinctly identifying and standardizing FITS output from yt.  Since it would not be in conflict with the FITS standard itself, I see no reason not to codify the choices with a document like this.

**Review by chummels — COMMENTED** [review]
In general, the layout of this proposal seems sounds.  A few questions, though.

Is this proposed yt FITS image standard meant to operate in parallel to the existing `save_as_dataset()` functionality for storing the sorts of objects you describe like slices, projections, covering grids, etc.?  What is the benefit of saving these objects as FITS files as opposed to as native yt formats from `save_as_dataset()`?

Will these standardized `YTFITSDataset` files be compatible with external FITS viewers like ds9?

If someone reads in an external `FITS` file without all of the YT dataset metadata, and then uses `save_as_dataset()`, will the default behavior be to create a new `YTFITSDataset`, or are there necessary steps that a user must make in order to define specific metadata required to save as a `YTFITSDataset` like units and such?  

I guess the main question I have is if the proposed frontend is going to be used in a generalized way that benefits others, or if it will just be very specific to @jzuhone 's workflow.  I believe @jzuhone it will be general, but just want to confirm is all.

**Comment by jzuhone** [comment]
Hi @chummels, I'll answer these questions in turn. 

> Is this proposed yt FITS image standard meant to operate in parallel to the existing save_as_dataset() functionality for storing the sorts of objects you describe like slices, projections, covering grids, etc.? What is the benefit of saving these objects as FITS files as opposed to as native yt formats from save_as_dataset()?

Let me just stress for clarification--the FITS image file writing machinery is _already_ in yt and has been for a number of years (see [here](https://yt-project.org/doc/visualizing/writing_fits_images.html)), in the `FITSImageData` class and its subclasses, `FITSSlice`, `FITSProjection`, etc. 

All this YTEP does is specify a standard for the metadata in the file headers to define units and coordinates. The main advantage over `save_as_dataset` is writing the projected, sliced, or gridded data to a format that other applications (such as ds9, glue, etc.) can read. This facilities comparisons to real observations.

> Will these standardized YTFITSDataset files be compatible with external FITS viewers like ds9?

They already are, yes.

> If someone reads in an external FITS file without all of the YT dataset metadata, and then uses save_as_dataset(), will the default behavior be to create a new YTFITSDataset, or are there necessary steps that a user must make in order to define specific metadata required to save as a YTFITSDataset like units and such?

This machinery doesn't interact with `save_as_dataset()` at all, currently, it is a separate thing. Currently, there are special subclasses of `FITSImageData`, such as `FITSSlice`, `FITSProjection`, etc, which will handle writing the metadata to the headers automatically.

I'd recommend looking at the section of the yt docs I linked to above and the PR https://github.com/yt-project/yt/pull/2010 for more details about how the FITS file I/O is already implemented in yt and also what changes I would make to this to enforce the standard proposed in the YTEP.

File: source/YTEPs/YTEP-0034.rst [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    CRPIX2  =                256.5
+    CDELT1  =            0.9765625
+    CDELT2  =            0.9765625
+    CUNIT1  = 'kpc     '
+    CUNIT2  = 'kpc     '
+    CTYPE1  = 'LINEAR  '
+    CTYPE2  = 'LINEAR  '
+    CRVAL1  =                  0.0
+    CRVAL2  =                  0.0
+    LATPOLE =                 90.0
```
Comment (Cadair) [inline source/YTEPs/YTEP-0034.rst:170]: Do you need `LATPOLE` in this linear header? or was it put in by astropy?

File: source/YTEPs/YTEP-0034.rst [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    SIMPLE  =                    T / conforms to FITS standard
+    BITPIX  =                  -64 / array data type
+    NAXIS   =                    2 / number of array dimensions
+    NAXIS1  =                  512
+    NAXIS2  =                  512
+    EXTEND  =                    T
+    EXTNAME = 'DENSITY '           / extension name
+    BTYPE   = 'density '
+    BUNIT   = 'g/cm**3 '
+    LUNIT   =                  1.0 / kpc
```
Comment (Cadair) [inline source/YTEPs/YTEP-0034.rst]: The FITS 4.0 standard says:

> If the units of the keyword value are specified in the comment of the header keyword, it is recommended that the units string be enclosed in square brackets (i.e., enclosed by ‘[’ and ‘]’) at the beginning of the comment field, separated from the slash (’/’) comment field delimiter by a single space character. An example, using a non-standard keyword, is `EXPTIME = 1200. / [s] exposure time in seconds`

So I would recommend putting the units in these comments in `[]`.

**Review by Cadair — COMMENTED** [review]
I love a good exercise in FITS standards, so when I saw this on the mailing list I thought I would drive by review it :grin: 

It looks very nice, can't really see any issues with it other than the two nits I have pointed out inline.

I would be interested to try out a file with the example header to see how nicely it plays with all the tools I have for FITS files. I can see this being really useful for some solar simulations as well if we can make it play nice with Astropy/SunPy.

File: source/YTEPs/YTEP-0034.rst (hunk 1 above)
Comment (jzuhone) [reply source/YTEPs/YTEP-0034.rst:170]: AstroPy put this in. Is there a way to remove it except manually?

File: source/YTEPs/YTEP-0034.rst (hunk 2 above)
Comment (jzuhone) [reply source/YTEPs/YTEP-0034.rst]: @Cadair thanks for this suggestion--I have implemented it in the code and I'll also do it here.

**Comment by jzuhone** [comment]
@chummels did you have any comments on my answers? want to make sure I've resolved your questions.

**Comment by brittonsmith** [comment]
I don't have any technical comments with respect to the yt FITS standard as I have relatively little FITS-related experience. However, in a general sense, I think this is a fantastic idea. Creating a standard for a `YTFITSDataset` has the dual benefit of making yt output readable by all of these external tools but also fully re-readable by yt itself with no loss of functionality in terms of unit systems, metadata, etc. I think this plays perfectly with the philosophy behind `save_as_dataset` and will be a great addition.

**Review by brittonsmith — APPROVED** [review]

File: source/YTEPs/YTEP-0034.rst (hunk 1 above)
Comment (Cadair) [reply source/YTEPs/YTEP-0034.rst:170]: probably not :roll_eyes:

File: source/YTEPs/YTEP-0034.rst [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+one containing a 2 or 3-dimensional array with will correspond to a "field" in
+standard yt parlance. The dimensions of each array shall be consistent with the
+others for the entire file. The first or "primary" HDU will also contain an 
+image array. 
+
+Header Information
+==================
+
+Each FITS header associated with a a field in the FITS file shall be 
+entirely self-descripting with respect to the properties of the field, the
```
Comment (chummels) [inline source/YTEPs/YTEP-0034.rst]: self-describing

**Comment by chummels** [comment]
Thank you for the clarification, @jzuhone .  I think having standard header information will benefit our interaction with FITS files moving forward.  

I have a few additional comments relevant to the FITS frontend, but I realize that they aren't directly applicable to the work proposed in this YTEP.  It just seems like here would be a good place to discuss them.

Right now, as you point out, the FITS functionality seems somewhat disconnected from many other similar parts of the code.  Maybe it is worthwhile to integrate them a bit more?  For example, as noted in the [docs you linked on FITS writing](https://yt-project.org/doc/visualizing/writing_fits_images.html), one can only save projections or slices as a fits file using a totally different class than the traditional `SlicePlot` or `ProjectionPlot`.  Do you think it would ever be possible to just throw a method on these classes to save as a fits file, i.e.,

```
# to see what it looks like
prj = yt.ProjectionPlot(ds, "z", ["temperature"], weight_field="density", width=(500., "kpc"))
prj.show()

# to save
prj.save_as_fits()
```

instead of the current workflow:

```
# to see what it looks like
prj = yt.ProjectionPlot(ds, "z", ["temperature"], weight_field="density", width=(500., "kpc"))
prj.show()

# to save
prj_fits = yt.FITSProjection(ds, "z", ["temperature"], weight_field="density")
prj_fits.writeto("sloshing.fits")
```

Or maybe this is unfeasible for technical reasons.  Just an idea...

**Review by chummels — APPROVED** [review]

**Comment by matthewturk** [comment]
@chummels One slightly lower-impact solution might be to implement a check in the `save` function that looks at the extension and if it's `.fits`, use that workflow.  I am not super keen on adding a `save_as_fits` type function, and I think we already do some checks on the extension to see which matplotlib canvas object to use, so it wouldn't be a big deal to add this as well.

**Comment by jzuhone** [comment]
Hi @chummels, 

Thanks for the good feedback! There are a couple of technical issues with adding this to `PlotWindow`, however. 

1. The resolution of the image is a slightly different concept for `PlotWindow` vs FITS images. In the former, we just use a big number (800 pixels on a side IIRC) to make sure the plot looks good visually and communicates information to the eye. I suspect this is why the option to change the image resolution is actually fairly well hidden in `PlotWindow`. But for FITS images, the image resolution might actually mean something physically--one might be trying to match the minimum resolution of an AMR dataset or trying to match the angular resolution of an instrument. In the `FITSImageData` subclasses which make slices and projections, setting the image resolution is something that can be done easily up front. 

2. I wouldn't want to write a FITS image directly to disk from a slice or a projection because one of the important aspects of `FITSImageData` is that it is basically a unit-aware wrapper around AstroPy's `HDUList` class and is designed to replicate some of the latter's functionality, in addition to some other methods I added. This means that a lot of its functionality is intended to be used in-memory before writing to disk.

**Comment by ngoldbaum** [comment]
>  I suspect this is why the option to change the image resolution is actually fairly well hidden in PlotWindow.

It has more to do with the fact that matplotlib lets you set the figure size and dpi but not really the resolution of the image.

**Comment by jzuhone** [comment]
I have to admit I am kind of warming up to @chummels' idea but I'd want it to be a separate method like `to_fits` and would want to be able to change the image resolution within that method.

**Comment by jzuhone** [comment]
Thanks to everyone for the comments and feedback--this is ready to go. Still working on the actual PR on the yt side.

File: source/YTEPs/YTEP-0034.rst (hunk 3 above)
Comment (ngoldbaum) [reply source/YTEPs/YTEP-0034.rst]: @jzuhone forgot this one

**Review by ngoldbaum — APPROVED** [review]
LGTM other than the spelling nit I commented about.

**Comment by jzuhone** [comment]
@ngoldbaum thanks for pointing that out, fixed.

**Pull request merged**